File: src/lib/parser/parser.ts
import { parseAliases } from "./aliases.ts"
import { parseCombos } from "./combos.ts"
import { parseLayer } from "./layer.ts"
import { parseOptions } from "./options.ts"
import { parseStructure } from "./structure.ts"
import { Block, Layout } from "../types.ts"
import { findBlocks, getOptionalBlock, getRequiredBlock } from "./blocks.ts"

export function parseLayoutFile(filename: string): Layout {
  const { blocks } = parseMarkdownFile(filename)
  return parseBlocks(blocks)
}

export function parseMarkdownFile(filename: string): { lines: string[]; blocks: Block[] } {
  const lines = Deno.readTextFileSync(filename).split("\n")
  const blocks = findBlocks(lines)

  return { lines, blocks }
}

export function parseBlocks(blocks: Block[]): Layout {
  const options = parseOptions(getOptionalBlock(blocks, "options")?.lines)
  const structure = parseStructure(getRequiredBlock(blocks, "structure"))
  const layers = blocks.filter((b) => b.name.startsWith("layer:")).map((block) => parseLayer(block, structure))

  const aliasesBlock = getOptionalBlock(blocks, "aliases")
  const aliases = aliasesBlock ? parseAliases(aliasesBlock) : undefined

  const combosBlock = getOptionalBlock(blocks, "combos")
  const combos = combosBlock ? parseCombos(combosBlock) : undefined

  return { options, structure, layers, aliases, combos }
}

File: internal/parser/parser.go
package parser

import (
	"strings"

	"github.com/elentok/qmkmd/internal/layout"
)

func ParseLayoutFile(filename string) (layout.Layout, error) {
	_, blocks, err := ParseMarkdownFile(filename)
	if err != nil {
		return layout.Layout{}, err
	}
	return ParseBlocks(blocks)
}

func ParseBlocks(blocks []layout.Block) (layout.Layout, error) {
	var optionsLines []string
	optionsBlock, err := GetOptionalBlock(blocks, "options")
	if err != nil {
		return layout.Layout{}, err
	}
	if optionsBlock != nil {
		optionsLines = optionsBlock.Lines
	}
	options := ParseOptions(optionsLines)

	structureBlock, err := GetRequiredBlock(blocks, "structure")
	if err != nil {
		return layout.Layout{}, err
	}
	structure, err := ParseStructure(structureBlock)
	if err != nil {
		return layout.Layout{}, err
	}

	layers := make([]layout.Layer, 0)
	for _, block := range blocks {
		if !strings.HasPrefix(block.Name, "layer:") {
			continue
		}
		layer, err := ParseLayer(block, structure)
		if err != nil {
			return layout.Layout{}, err
		}
		layers = append(layers, layer)
	}

	var aliases map[string]layout.Alias
	aliasesBlock, err := GetOptionalBlock(blocks, "aliases")
	if err != nil {
		return layout.Layout{}, err
	}
	if aliasesBlock != nil {
		aliases = ParseAliases(*aliasesBlock)
	}

	var combos []layout.Combo
	combosBlock, err := GetOptionalBlock(blocks, "combos")
	if err != nil {
		return layout.Layout{}, err
	}
	if combosBlock != nil {
		combos = ParseCombos(*combosBlock)
	}

	return layout.Layout{
		Options:   options,
		Structure: structure,
		Layers:    layers,
		Aliases:   aliases,
		Combos:    combos,
	}, nil
}
